lib/main.py

```diff
@@ -27,57 +27,54 @@ async def lifespan(_app: FastAPI):
 
 
 APP = FastAPI(lifespan=lifespan)
-TASK_LIST: queue.Queue = queue.Queue(maxsize=100)
 
 
 class BackgroundProcessTask(threading.Thread):
     def run(self, *args, **kwargs):  # pylint: disable=unused-argument
+        nc = NextcloudApp()
+
+        provider_ids = set()
+        task_type_ids = set()
+        for chain_name, _ in chains.items():
+            provider_ids.add("llm2:" + chain_name)
+            (model, task) = chain_name.split(":", 2)
+            task_type_ids.add("core:text2text:" + task)
+
         while True:
-            task = TASK_LIST.get(block=True)
+            response = nc.providers.task_processing.next_task(list(provider_ids), list(task_type_ids))
+            if not isinstance(response, dict):
+                time.sleep(5)
+                continue
+
+            task = response["task"]
+            provider = response["provider"]
+
             try:
-                chain_name = task.get("chain")
+                chain_name = provider["name"][5:]
                 print(f"chain: {chain_name}", flush=True)
                 chain_load = chains.get(chain_name)
                 if chain_load is None:
-                    NextcloudApp().providers.text_processing.report_result(
-                        task["id"], error="Requested model is not available"
+                    NextcloudApp().providers.task_processing.report_result(
+                        task["id"], error_message="Requested model is not available"
                     )
                     continue
                 chain = chain_load()
                 print("Generating reply", flush=True)
                 time_start = perf_counter()
-                print(task.get("prompt"))
-                result = chain.invoke(task.get("prompt")).get("text")
+                print(task.get("input").get("input"))
```

Comment:
```python3
task.get("input").get("input")
````

if first `get` will return None, then next `get` will fail with exception anyway?

```python3
task.get("input", {}).get("input")
```

maybe this should be done this way if can not assume that first `input` is always present? Does we?

Reply:
I don't see a point in catching these errors, as they simply indicate errors on the server side and there is no way to actually recover from these errors. The chaining doesn't concern me either, as the first one will already fail before the second one can even happen.